GEODE-9365: Removing unnecessary synchronization

mhansonp opened the pull request:
Plus some cleanup

jchen21 commented on geode-core/src/main/java/org/apache/geode/internal/cache/ha/HARegionQueue.java:
```diff
@@ -2292,66 +2270,79 @@ public void destroy() throws CacheWriterException {
+      long startTime = System.currentTimeMillis();
+      synchronized (this.permitMon) {
+
+        if (putPermits.decrementAndGet() >= 0) {
```
getAndDecrement()?

mhansonp replied:
Nope that is correct.

mhansonp commented on geode-core/src/integrationTest/java/org/apache/geode/internal/cache/ha/BlockingHARegionJUnitTest.java:
```diff
@@ -281,26 +231,15 @@ public void testConcurrentPutsTakesNotExceedingLimit() throws Exception {
-      }
-    };
-    GeodeAwaitility.await().untilAsserted(ev);
+    assertThat(thread1.isAlive()).isTrue();
```
The test is disabled. I chose not to modify it.

mhansonp commented on geode-core/src/main/java/org/apache/geode/internal/cache/ha/HARegionQueue.java:
```diff
@@ -2292,66 +2270,79 @@ public void destroy() throws CacheWriterException {
      * in the HARegionQueue.
      */
     @Override
-    @edu.umd.cs.findbugs.annotations.SuppressWarnings("TLW_TWO_LOCK_WAIT")
```
This is the main body of the change.

mhansonp commented on geode-core/src/main/java/org/apache/geode/internal/cache/ha/HARegionQueue.java:
```diff
@@ -2292,66 +2270,79 @@ public void destroy() throws CacheWriterException {
      * @return int current Put permits
      */
     private int reconcilePutPermits() {
-      putPermits += takeSidePutPermits;
```
additional pieces required.

mhansonp commented on geode-core/src/main/java/org/apache/geode/internal/cache/ha/HARegionQueue.java:
```diff
@@ -2211,22 +2194,16 @@ public Object updateHAEventWrapper(InternalDistributedMember sender,
      * Current put permits available
      */
-    private int putPermits;
+    private final AtomicInteger putPermits = new AtomicInteger(0);
```
Variable changes that were required.

kirklund reviewed (changes requested):
I think you should retrofit the test to use `ExecutorServiceRule` (geode-junit) and `ErrorCollector` (JUnit) instead.

Change all the catch-blocks in threads to hand off the error to the `ErrorCollector` instead of appending to a `StringBuffer`:
```
@Rule
public ErrorCollector errorCollector = new ErrorCollector();
```
```
} catch (Exception e) {
  errorCollector.addError(e);
}
```
And replace all the `Thread` uses with simple uses of `ExecutorServiceRule`.

mhansonp commented:
> I think you should retrofit the test to use `ExecutorServiceRule` (geode-junit) and `ErrorCollector` (JUnit) instead.
> 
> Change all the catch-blocks in threads to hand off the error to the `ErrorCollector` instead of appending to a `StringBuffer`:
> 
> ```
> @Rule
> public ErrorCollector errorCollector = new ErrorCollector();
> ```
> 
>  ```
> } catch (Exception e) {
>   errorCollector.addError(e);
> }
> ```
> 
> And replace all the `Thread` uses with simple uses of `ExecutorServiceRule`.
> 
> Submitting tasks to `ExecutorServiceRule` will return futures the test can wait on (with built in use of GeodeAwaitility timeout) so you can ditch the waiting, any looping recommended by others, etc.

I could use some guidance on this. I will talk to you directly.

kirklund reviewed (approved):
Test needs more love at a later date.

jchen21 commented on geode-core/src/main/java/org/apache/geode/internal/cache/ha/HARegionQueue.java:
```diff
@@ -2292,66 +2270,79 @@ public void destroy() throws CacheWriterException {
+
+        long duration = (System.currentTimeMillis() - startTime);
+
+        if (reconcilePutPermits() >= 0) {
```
I think this should be `if (reconcilePutPermits() > 0) {` without the `=`.  Because the original code that you deleted is `if (reconcilePutPermits() <= 0) {`.

dschneider-pivotal replied:
I think the new code is different here because this thread already decremented putPermits and it went negative. In the old code we waited until putPermits was > 0 BEFORE we decremented it. But now we decrement it even if it is already < 0. 
reconcilePutPermits  updates and returns putPermits so if we see it go positive (>= 0) then this thread now has a permit to add to the queue.

nabarunnag replied:
just trying to understand the change, previously we alway decremented the putPermits, now we don't allow the block to be executed if the putPermits is zero after decrement but we allowed it earlier.

mhansonp commented:
Updated with a suggested performance change.

DonalEvans reviewed (approved):
Approving the changes in `HARegion`, which is the only file in the PR for which I'm codeowner.

nabarunnag commented on geode-core/src/main/java/org/apache/geode/internal/cache/ha/HARegionQueue.java:
```diff
@@ -2292,66 +2270,79 @@ public void destroy() throws CacheWriterException {
+        throw new InterruptedException();
+      }
+
+      if (putPermits.decrementAndGet() >= 0) {
```
Was wondering on the behavior change for value 0, previously if the value was zero, we allowed the code block to execute. Is it ok in the new code not to?

nabarunnag commented:
Is there a behavior change in this code ? just wondering if there is a need for a test to be written?

nabarunnag commented:
also please do write a better commit message during the final commit explaining what has changed : finals added , refactored generics, cleanups of dead comments, new atomic counters.

mhansonp commented:
> also please do write a better commit message during the final commit explaining what has changed : finals added , refactored generics, cleanups of dead comments, new atomic counters.

I kind of pointed that stuff out in the comments I added when I created this PR, but point taken. I will do a better job of that in the future.

agingade commented:
It is about the number of items to keep in the HARegion and then throttle; which can go beyond set limit. 
Can we just replace the whole code using capacity and region’s size? and keep it simple...
sync(this) {
  if (capacity - region.size()) > 0) {
    wait(x time); // Notify when take is done while still waiting...
   }
}
